Makefile

```diff
@@ -1,7 +1,17 @@
 .PHONY: test
 test:
 	python -m pip install mock
-	for file in $(shell find . -name test -type d); do python -m unittest discover -v -p '*_test.py' -s $$(dirname $$file); done
+	@failed=0; \
+    for file in $(shell find . -name test -type d); do \
+        echo "Running tests in $$(dirname $$file)"; \
+        python -m unittest discover -v -p '*_test.py' -s $$(dirname $$file) || failed=1; \
+    done; \
+    if [ $$failed -eq 0 ]; then \
+        echo "All tests passed."; \
+    else \
+        echo "One or more tests have failed."; \
+        exit 1; \
+    fi
```

Comment:
Thanks for the fix! Maybe it makes sense to put this into a separate `test.sh` script as it's syntactically getting hard to maintain in the Makefile target. Maybe there is also a way to achieve this through a Python script where you trigger `unittest discover` programmatically: https://www.doc.ic.ac.uk/~nuric/coding/how-to-discover-and-run-unit-tests-programmatically-in-python.html

Reply:
@Gerrit91 I tried doing it with python, which could work. But there were many import problem I was not able to solve in reasonable time.

Maybe you have an idea to fix the imports? Or we use the shell for the moment. Here is my python script:
```
#!/usr/bin/env python
import os
import unittest

TEST_FOLDER_NAME = "test"

def discover_and_run(start_dir: str = '.', pattern: str = '*_test.py'):
    """Discover and run tests cases, returning the result."""
    test_loader = unittest.TestLoader()
    runner = unittest.TextTestRunner()
    
    for root, dirs, _ in os.walk(start_dir):
        if TEST_FOLDER_NAME in dirs:
            test_directory = os.path.join(root, TEST_FOLDER_NAME)
            print("Running tests in: " + test_directory)
            test_suite = test_loader.discover(start_dir=test_directory, pattern='*_test.py')
            runner.run(test_suite)

discover_and_run()

```

Reply:
It's okay for me with the bash script when it is not solvable through Python in a reasonable amount of time.